pyoozie/builder.py

```diff
@@ -27,56 +28,77 @@ def _coordinator_submission_xml(username, coord_xml_path, configuration=None, in
+        self.__transformations = []
+        self.__transformation_kwargs = {}
+
+    def add_action(self, name, action, depends_upon=None, **kwargs):
+        def create_action_name(name):
+            return tags.validate_xml_id('action-' + name)
+
+        # Validate name
+        name = create_action_name(name)
+        assert name not in self.__actions, "Cannot add an action with the same name (%(name)s) twice".format(
+            name=name)
+
+        # Compose and validate dependency names
+        dependencies = set()
+        for dependency_name in depends_upon or {}:
+            dependencies.add(create_action_name(dependency_name))
+        self.__dependencies[name] = dependencies
+
+        # Store actions
+        self.__actions[name] = action
+        self.__action_kwargs[name] = kwargs
 
         return self
 
+    def add_transform(self, fnc_transform, **kwargs):
+        assert fnc_transform not in self.__transformation_kwargs, (
+            "Cannot add a transformation function (%(fnc)r) twice".format(fnc=fnc_transform))
+        self.__transformation_kwargs[fnc_transform] = kwargs
+        self.__transformations.append(fnc_transform)
+        return self
+
     def build(self, indent=False):
-        def format_xml(xml):
-            xml = xml.replace("<?xml version='1.0' encoding='UTF-8'?>", '')
-            return '\n'.join([(' ' * 8) + line for line in xml.strip().split('\n')])
-        return '''
-<?xml version="1.0" encoding="UTF-8"?>
-<workflow-app xmlns="uri:oozie:workflow:0.5"
-              name="{name}">
-    <start to="{action_name}" />
-    <action name="{action_name}">
-{action_payload_xml}
-        <ok to="end" />
-        <error to="action-error" />
-    </action>
-    <action name="action-error">
-{action_error_xml}
-        <ok to="kill" />
-        <error to="kill" />
-    </action>
-    <kill name="kill">
-        <message>{kill_message}</message>
-    </kill>
-    <end name="end" />
-</workflow-app>
-'''.format(action_payload_xml=format_xml(self._action_payload.xml(indent=indent)),
-           action_error_xml=format_xml(self._action_error.xml(indent=indent)),
-           kill_message=self._kill_message,
-           action_name=self._action_name,
-           name=self._name).strip()
+        env = {
+            'actions': self.__actions,
+            'action_kwargs': self.__action_kwargs,
+            'dependencies': self.__dependencies,
+            'workflow': self.__workflow,
+        }
+
+        # Ensure that this operation can only be called once
+        assert self.__workflow, "Workflow already built"
+        self.__workflow = None
+
+        # Layout the action nodes provided
+        env = self.__action_layout_strategy(env)
+
+        # Manipulate those action nodes (and potentially add others)
+        for transformation in self.__transformations:
+            transformation_kwargs = self.__transformation_kwargs[transformation]
+            env = transformation(env, **transformation_kwargs)
```

Comment:
Apply transformations in order to add features such as:
  - A final success notification (`transforms.final_action`); and
  - Notifications actions on error (`transforms. add_actions_on_error`).